File: articles/what-is-biology-life-evolution-living-systems/r/biodiversity_summary.R
# Biodiversity summary workflow in R.

counts_path <- file.path("data", "biodiversity_counts.csv")
if (!file.exists(counts_path)) {
  counts_path <- file.path("..", "data", "biodiversity_counts.csv")
}

counts <- read.csv(counts_path)
rownames(counts) <- counts$site
counts$site <- NULL

shannon <- apply(counts, 1, function(x) {
  p <- x / sum(x)
  -sum(p[p > 0] * log(p[p > 0]))
})

summary_df <- data.frame(
  site = rownames(counts),
  richness = rowSums(counts > 0),
  total_abundance = rowSums(counts),
  shannon_diversity = shannon
)

print(round(summary_df, 4))

File: articles/what-is-biology-life-evolution-living-systems/python/biodiversity_summary.py
"""
Biodiversity summary workflow.

Run:
    python python/biodiversity_summary.py
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd

from biology_core import shannon_diversity


ARTICLE_DIR = Path(__file__).resolve().parents[1]
BIODIV_PATH = ARTICLE_DIR / "data" / "biodiversity_counts.csv"


def main() -> None:
    counts = pd.read_csv(BIODIV_PATH).set_index("site")
    taxon_cols = list(counts.columns)

    rows = []

    for site, row in counts.iterrows():
        rows.append(
            {
                "site": site,
                "richness": int((row[taxon_cols] > 0).sum()),
                "total_abundance": float(row[taxon_cols].sum()),
                "shannon_diversity": shannon_diversity(row[taxon_cols]),
            }
        )

    print(pd.DataFrame(rows).round(4).to_string(index=False))


if __name__ == "__main__":
    main()
